### ISSUE TITLE
incorrect geolocation

### ISSUE THREAD

**Issue by tansaku** [AgileVentures/WebsiteOne#1239]
@kenshinji reports his geolocation is incorrect:

> I don’t know why my profile displayed on agileventures.org showing my place is in US, and I can’t find the way to update it to China. http://www.agileventures.org/users/zeng-fred-wang

I ran the job specified in #1227 thinking it might update the location, but it didn't, but doh! that's only updating those who don't have geocodes - am rerunning the job to update all users geolocation

**Comment by tansaku** [comment]
okay, so even re-running the job for every user did not seem to have any impact on @kenshinji's geolocation ...

**Comment by kenshinji** [comment]
@tansaku thanks, so the geolocation is get from the users when they login into agileventures.org? And there is no way users can update it manually?

**Comment by tansaku** [comment]
@kenshinji I'm not sure, I didn't develop the system - looking at the user model I see:

```rb  
after_validation :geocode, if: ->(obj) { obj.last_sign_in_ip_changed? }
```

so I'm thinking that it's currently set to work out geolocation based on the ip address that the browser receives

**Comment by tansaku** [comment]
I think it's from this gem https://github.com/alexreisner/geocoder and `last_sign_in_ip_changed` is a reference to [Devise](http://www.rubydoc.info/github/plataformatec/devise/Devise/Models/Trackable) outside of its usage in our codebase ... perhaps your ISP makes us think that your in the US?  from the db I see your current_sign_in_ip: "116.192.8.195", but that maps to China when I check independently.  

Our geocoder config is set to look up via `freegeoip` so maybe there's a problem there ...

We could add a feature to allow the user to override ...  I could also just override specifically in the db for you ...

**Comment by tansaku** [comment]
Ah, seems like there is a conflict between your last sign in ip and your current sign in ip: 
```
irb(main):007:0> Geocoder.search "192.241.202.248"
=> [#<Geocoder::Result::Freegeoip:0x007fdf9f4a4418 @data={"ip"=>"192.241.202.248", "country_code"=>"US", "country_name"=>"United States", "region_code"=>"CA", "region_name"=>"California", "city"=>"San Francisco", "zip_code"=>"94119", "time_zone"=>"America/Los_Angeles", "latitude"=>37.7749, "longitude"=>-122.4194, "metro_code"=>807}, @cache_hit=nil>]
irb(main):008:0> Geocoder.search "116.192.8.195"
=> [#<Geocoder::Result::Freegeoip:0x007fdf9f44f0f8 @data={"ip"=>"116.192.8.195", "country_code"=>"CN", "country_name"=>"China", "region_code"=>"31", "region_name"=>"Shanghai", "city"=>"Shanghai", "zip_code"=>"", "time_zone"=>"Asia/Shanghai", "latitude"=>31.0456, "longitude"=>121.3997, "metro_code"=>0}, @cache_hit=nil>]
```

**Comment by tansaku** [comment]
one is US and the other is China ... our model is checking for changes in the last_sign_in_ip:

```rb
  geocoded_by :last_sign_in_ip do |user, results|
    if geo = results.first
      user.latitude = geo.data['latitude']
      user.longitude = geo.data['longitude']
      user.city = geo.data['city']
      user.region = geo.data['region_name']
      user.country_name = geo.data['country_name']
      user.country_code = geo.data['country_code']
    end
  end
```

I manually override your last_sign_in_ip to be your current_sign_in_ip and saved you and re-ran the geocoding and I think you'll find you're updated in the AV site now ...

I think maybe we need to update our system so that it determines location based on current_sign_in_ip rather than last_sign_in_ip unless @tochman @yggie or @diraulo can tell me there's some reason to prefer last_sign_in_ip ...

**Comment by kenshinji** [comment]
Yeah, that makes sense, because I browser the Internet through a proxy, in order to access those websites which are blocked by GFW.

**Comment by tansaku** [comment]
GFW = Google Fire Wall?

**Comment by kenshinji** [comment]
It is [Great Firewall of China](https://en.wikipedia.org/wiki/Great_Firewall) , which blocks Google, Youtube, Twitter, etc... And the list goes on and on :(
